core/domain/exp_domain.py

```diff
@@ -2514,12 +2515,1064 @@ def _convert_states_v50_dict_to_v51_dict(
+                        if rule_spec['rule_type'] == 'IsInclusivelyBetween':
+                            if (
+                                rule_spec['inputs']['a'] > rule_spec[
+                                    'inputs']['b']
+                            ):
+                                rule_spec['inputs']['a'], rule_spec[
+                                    'inputs']['b'] = rule_spec['inputs'][
+                                        'b'], rule_spec['inputs']['a']
+
+                state_dict['interaction']['answer_groups'] = answer_groups
+
+            # FractionInput Interaction.
+            if state_dict['interaction']['id'] == 'FractionInput':
+                # All rules should have solutions that do not match
+                # previous rules' solutions.
+                answer_groups = state_dict['interaction']['answer_groups']
+                answer_groups = cls._fraction_rules_should_not_match_prev_rule(
+                    answer_groups
+                )
+                state_dict['interaction']['answer_groups'] = answer_groups
+
+            # MultipleChoiceInput Interaction.
+            if state_dict['interaction']['id'] == 'MultipleChoiceInput':
+                selected_equals_choices = []
+                answer_groups = state_dict['interaction']['answer_groups']
+                # No answer choice should appear in more than one answer group.
+                for answer_group in answer_groups:
+                    unwanted_rule = []
+                    for rule_spec in answer_group['rule_specs']:
+                        if rule_spec['rule_type'] == 'Equals':
+                            if (
+                                rule_spec['inputs']['x'] in
+                                selected_equals_choices
+                            ):
+                                unwanted_rule.append(rule_spec)
+                            else:
+                                selected_equals_choices.append(
+                                    rule_spec['inputs']['x'])
+                    for ele in unwanted_rule:
+                        answer_group['rule_specs'].remove(ele)
+                    # Removal of answer group if it is empty check, investigation part.
+                    if len(answer_group['rule_specs']) == 0:
+                        answer_groups.remove(answer_group)
+
+                # Answer choices should be non-empty and unique.
+                choices = (
+                    state_dict['interaction']['customization_args'][
+                        'choices']['value']
+                )
+
+                choices, answer_groups = (
+                    cls._choices_should_be_unique_and_non_empty(
+                        choices, answer_groups)
+                )
+                state_dict['interaction']['customization_args']['choices'][
+                    'value'] = choices
+
+                state_dict['interaction']['answer_groups'] = answer_groups
```

Comment:
Do we need these lines?

Reply:
I think so, I just updated in case. I will check this part and get back to you.